<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>(m,n)-mer - A Simple and New Statistical Feature for viral Classification â€¢ mnmer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="(m,n)-mer - A Simple and New Statistical Feature for viral Classification">
<meta property="og:description" content="The (m,n)-mer is a new statistical feature based upon conditional frequencies (conditional probability density distributions). It reads a FASTA file and generates a table describing the conditional frequency distribution of the selected (m,n)-mer in the sequences. This output is combined with class information to generate the feature matrix for classification. The mn function generates the feature matrix using conditional probability from the datasets. This function can generate both k-mers and mn-mers.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mnmer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette.html">(m,n)-mer - A Simple and New Statistical Feature for viral Classification</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="https://img.shields.io/github/license/labinfo-lncc/mnmer?style=plastic" alt="GitHub"><img src="https://img.shields.io/github/repo-size/labinfo-lncc/mnmer?style=plastic" alt="GitHub repo size"><img src="https://img.shields.io/github/r-package/v/labinfo-lncc/mnmer?style=plastic" alt="GitHub R package version"><!-- badges: start --><a href="https://github.com/labinfo-lncc/mnmer/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/labinfo-lncc/mnmer/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a> <!-- badges: end --></p>
<div class="figure">
<img src="https://user-images.githubusercontent.com/57667417/191082345-57fed066-37e9-4a8a-a65a-c9562d0625a4.png" alt=""><p class="caption">logo</p>
</div>
<div class="section level1">
<div class="page-header"><h1 id="conditional-frequency-distribution">Conditional frequency distribution<a class="anchor" aria-label="anchor" href="#conditional-frequency-distribution"></a>
</h1></div>
<p>The (m,n)-mer R package was created to summarize biological data into numerical characteristics. It reads a FASTA file and generates a table describing the conditional frequency distribution of the selected (m,n)-mer in the sequences. This output is combined with class information to generate the feature matrix for classification.</p>
<p>(m,n)-mers are an alternative for k-mers (Figure 1). We proposed the replacement of the unconditional k-mer frequency by the conditional frequency, which represents the relative frequency of the n-mer conditioned to f m-mer that precedes it. For more details and performance comparison, please see Andrade et al., 2022 (in press).</p>
<div class="figure">
<img src="https://user-images.githubusercontent.com/57667417/191081859-0b0ae464-f257-4c82-9dea-8d4629605357.png" alt=""><p class="caption">Fig1</p>
</div>
<p><strong>Fig 1.</strong> Comparing k-mer to mn-mer relative frequency.</p>
<p>According to Figure 2, the k-mers are represented as (0,k) and the mn-mers as (m,n).</p>
<div class="figure">
<img src="https://user-images.githubusercontent.com/57667417/193356992-faa03a21-fb0a-48cc-b3fb-975021060b79.png" alt=""><p class="caption">Fig2</p>
</div>
<p><strong>Fig 2.</strong> Numeric representation.</p>
<p>The output table (Figure 3) includes the fasta file accession numbers as an ID column, the relative frequency of mn-mers up to 4^k columns, and class information.</p>
<p><img src="https://user-images.githubusercontent.com/57667417/191082016-b6835c4c-c115-498d-a2d1-c7d93ec20fe5.png" alt="Fig3"><strong>Fig 3.</strong> Output example.</p>
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<pre><code>R version 4.0.0 or later
Biostrings
Utils</code></pre>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"labinfo-lncc/mnmer"</span>, ref<span class="op">=</span><span class="st">"main"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="quick-start-running-mn-mer-on-example-dataset">Quick Start: Running (m,n)-mer on example dataset<a class="anchor" aria-label="anchor" href="#quick-start-running-mn-mer-on-example-dataset"></a>
</h2>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"mnmer"</span><span class="op">)</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"mnmer"</span><span class="op">)</span></span></code></pre>
<p>Assume we need to distinguish between viruses detected in mosquito samples and viruses that exclusively infect plants. The mn function generates the feature matrix using conditional probability from the datasets mosquito vir.fasta and plant vir.fasta. This function can generate both k-mers and mn-mers.</p>
<div class="section level4">
<h4 id="producing-k-mers">Producing k-mers<a class="anchor" aria-label="anchor" href="#producing-k-mers"></a>
</h4>
<p>The parameter <code>k</code> is set to choice for k-mer generation, while the parameter <code>m</code> is set to zero. Considering that the k-mers are conditioned to zero bases.</p>
<pre><code><span><span class="va">mosquito</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"mosquito_vir.fasta.gz"</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">plant</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"plant_vir.fasta.gz"</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="producing-mn-mers">Producing (m,n)-mers<a class="anchor" aria-label="anchor" href="#producing-mn-mers"></a>
</h4>
<p>The <code>k</code> and <code>m</code> parameters are chosen by the user for mn-mer creation. For instance, <code>k = 2</code> and <code>m = 1</code> yield the (1,1)-mer, in which one base is conditioned on the frequency of one preceding base.</p>
<pre><code><span><span class="va">mosquito</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"mosquito_vir.fasta"</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">plant</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"plant_vir.fasta"</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
<p>All sequences with N + IUPAC bases content higher than 10% are removed from further analysis.</p>
<p>For classification outside of the mnmer program, we utilize the (1,1)-mer feature matrices. Hereâ€™s a real-world example of code:</p>
<pre><code><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MLeval</span><span class="op">)</span></span>
<span></span>
<span><span class="va">classes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mosquito</span><span class="op">)</span>, <span class="st">"mosquito.vir"</span><span class="op">)</span></span>
<span><span class="va">featureMatrix_mosquito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">mosquito</span>,<span class="va">classes</span><span class="op">)</span></span>
<span><span class="va">classes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">plant</span><span class="op">)</span>, <span class="st">"plant.vir"</span><span class="op">)</span></span>
<span><span class="va">featureMatrix_plant</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">plant</span>,<span class="va">classes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">featureMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">featureMatrix_mosquito</span>, <span class="va">featureMatrix_plant</span><span class="op">)</span></span>
<span><span class="va">featureMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">featureMatrix</span>, select <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">seqid</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">train_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/createDataPartition.html" class="external-link">createDataPartition</a></span><span class="op">(</span><span class="va">featureMatrix</span><span class="op">$</span><span class="va">classes</span>, p<span class="op">=</span><span class="fl">0.8</span>, list<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">train</span> <span class="op">&lt;-</span> <span class="va">featureMatrix</span><span class="op">[</span><span class="va">train_index</span>, <span class="op">]</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">featureMatrix</span><span class="op">[</span><span class="op">-</span><span class="va">train_index</span>, <span class="op">]</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/trainControl.html" class="external-link">trainControl</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"cv"</span>, summaryFunction<span class="op">=</span><span class="va">twoClassSummary</span>, classProbs<span class="op">=</span><span class="cn">T</span>, savePredictions <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">roc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/train.html" class="external-link">train</a></span><span class="op">(</span><span class="va">classes</span> <span class="op">~</span> <span class="va">.</span>, data<span class="op">=</span><span class="va">train</span>, method<span class="op">=</span><span class="st">"rf"</span>, preProc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span><span class="op">)</span>, trControl<span class="op">=</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MLeval/man/evalm.html" class="external-link">evalm</a></span><span class="op">(</span><span class="va">roc</span><span class="op">)</span> <span class="co"># Make the ROC plot</span></span></code></pre>
<p>This classification produces the ROC curve and metrics shown below:</p>
<div class="figure">
<img src="https://user-images.githubusercontent.com/57667417/191288837-2f13cee0-96f8-48fb-a4e0-e7e28d832efe.png" alt=""><p class="caption">Rplot01</p>
</div>
<table class="table">
<thead><tr class="header">
<th>Metrics</th>
<th>Value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AUC</td>
<td>1.0000</td>
</tr>
<tr class="even">
<td>ROC</td>
<td>1.0000</td>
</tr>
<tr class="odd">
<td>Sensibility</td>
<td>0.9500</td>
</tr>
<tr class="even">
<td>Specificity</td>
<td>1.0000</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>All data used in this project are stored at: <a href="https://biotools.labinfo.lncc.br/mnmer/" class="external-link uri">https://biotools.labinfo.lncc.br/mnmer/</a></p>
</div>
<div class="section level2">
<h2 id="work-in-progress">Work in progress<a class="anchor" aria-label="anchor" href="#work-in-progress"></a>
</h2>
<p>If you have any queries or find a bug, please submit an issue on GitHub or email <a href="mailto:atrv@lncc.br" class="email">atrv@lncc.br</a>.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mnmer</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Amanda Andrade <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-3530-7435" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Marco Grivet <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-5130-095X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Otavio Brustolini <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-8132-9753" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Ana Vasconcelos <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0002-4632-2086" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Amanda Andrade, Marco Grivet, Otavio Brustolini, Ana Vasconcelos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
