<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>(m,n)-mer - A Simple Statistical Feature for Sequence Classification • mnmer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="(m,n)-mer - A Simple Statistical Feature for Sequence Classification">
<meta property="og:description" content="mnmer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mnmer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">(m,n)-mer - A Simple Statistical Feature for Sequence Classification</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>(m,n)-mer - A Simple Statistical Feature for
Sequence Classification</h1>
                        <h4 data-toc-skip class="author">Amanda Araújo
Serrão de Andrade</h4>
            <address class="author_afil">
      National Laboratory for Scientific Computing, Bioinformatics
Laboratory (LABINFO), Petrópolis, Rio de Janeiro,
Brazil<br><h4 data-toc-skip class="author">Marco
Grivet</h4>
            <address class="author_afil">
      Pontifícia Universidade Católica do Rio de Janeiro, 22451-900,
Brazil<br><h4 data-toc-skip class="author">Otávio
Brustolini</h4>
            <address class="author_afil">
      National Laboratory for Scientific Computing, Bioinformatics
Laboratory (LABINFO), Petrópolis, Rio de Janeiro,
Brazil<br><h4 data-toc-skip class="author">Ana
Tereza Ribeiro de Vasconcelos</h4>
            <address class="author_afil">
      National Laboratory for Scientific Computing, Bioinformatics
Laboratory (LABINFO), Petrópolis, Rio de Janeiro,
Brazil<br><div class="hidden name"><code>vignette.Rmd</code></div>

    </address>
</address>
</address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The (m,n)-mer is a new statistical feature based upon
      conditional frequencies (conditional probability density
      distributions). Here, we introduce the readNumFASTA and mnmer
      function, demonstrating its practical application in supervised
      classification.</p>
    </div>
    
<p>The (m,n)-mer is a k-mer variant that expresses conditional
probabilities instead of unconditional ones (Figure 1). Our proposal
suggests using the relative frequency of an n-mer conditioned to its
preceding m-mer, rather than the unconditional k-mer frequency. When
m+n=k, this new feature has an identical size and range to k-mer.</p>
<p>Here, we consider k-mers as an particular case of (m,n)-mers, where
k-mers are specifically (0,k)-mers (Figure 2). All other (m,n)-mers that
share the same size and range as a k-mer can be categorized as follows:
(1,k-1), (2,k-2), up to (k-1,1). For instance, the following three m-n
mers: (1,3)-mer, (2,2)-mer and (3, 1)-mer all have the same
characteristics as their related k-mer, which is in this case the
(0,4)-mer.</p>
<p>For more methodological details and feature performance comparison in
binary, multiclass and clustering classifications, please see Andrade et
al., 2022 (in press). </p>
<p>The R package is used to summarize biological data by generating a
table that shows the conditional frequency distribution of selected
(m,n)-mers in FASTA files. This is an alternative to using k-mers for
numerical characterization. By combining this output with class
information, a feature matrix suitable for classification is
created.</p>
<p>The output table (Figure 4) includes the FASTA file accession numbers
as an ID column, the relative frequency of (m,n)-mers up to <span class="math inline">\(4.k\)</span> columns, and class information.</p>
<p>The package needs R(&gt;= 4.2.0), Biostrings(&gt;= 3.1) and
Utils(&gt;= 2.0.0).</p>
<p>The user should install the package from the GitHub repository. It
can be done by using the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"labinfo-lncc/mnmer"</span>, ref<span class="op">=</span><span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
<p>This function employs Biostrings for reading FASTA files into the R
system. It enables users to limit the number of sequences loaded, select
sequences at random, and set a non-ACTG base cutoff percentage.</p>
<p>The parameters are:</p>
<p><code>FASTAfile</code> = It could be a multiFASTA.</p>
<p><code>size</code> = Number of sequences to be loaded.</p>
<p><code>rand</code> = Select sequences randomly or not. Set TRUE or
FALSE</p>
<p><code>pni</code> = Percentage of non-ACTG (default = 0.20)</p>
<p>As default, all sequences more than 20% of N + IUPAC bases will be
removed from further analysis given the little informative nature of
those bases. In that case, the <code>mnmer</code> function prints the
following warning:</p>
<pre><code><span><span class="co">## [1] "Warning: Sequence has a proportion of N + IUPAC bases = 30%"</span></span></code></pre>
<p>If the user would like to accept these sequences, the
<code>pni</code> parameter should be set to 0.00</p>
<p>The <code>readNumFASTA</code> function returns an DNAStringSet data
structure, used by the <code>mnmer</code> in further analysis. To learn
more about DNAStringSet please check Biostrings documentation.</p>
<p>The main function of this package is the <code>mnmer</code> function.
It creates dataframes with the conditional probability. By invoking the
function <code>cmmer</code> from the C++ script, this function can
create both k-mers and (m,n)-mers.</p>
<p>The parameters receives:</p>
<p><code>seqs</code> = DNAStringSet object</p>
<p><code>m</code> = Value of k for k-mer generation. Needs to be
different from zero.</p>
<p><code>n</code> = Value of m for (m,n)-mer generation in the format of
(m, k-m). In case of k-mer generation, m should be zero as (0,k).</p>
<p>Assume we need to classify viruses that infect human from viruses
that infect plants. The corresponding FASTA files can be found in the
extdata folder.</p>
<p>After package installation, the user should run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"mnmer"</span><span class="op">)</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"mnmer"</span><span class="op">)</span></span></code></pre></div>
<p>From the <code>readNumFASTA</code> function we load biological
sequences into R system as an DNAStringSet structure. In our practical
example, we would like to load 1,000 random sequences per FASTA. These
sequences must have less than 50% of non-ACTG bases. We should run:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/readNumFASTA.html">readNumFASTA</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"human_vir.fasta.gz"</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1000</span>,<span class="cn">TRUE</span>,<span class="fl">0.50</span><span class="op">)</span></span>
<span><span class="va">plant</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/readNumFASTA.html">readNumFASTA</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"plant_vir.fasta.gz"</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1000</span>,<span class="cn">TRUE</span>,<span class="fl">0.50</span><span class="op">)</span></span></code></pre></div>
<p>For k-mer generation, the parameter k is set to choice, while the
parameter m is set to zero. Given that the k-mers have been conditioned
to zero bases.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_02mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="va">human</span>,<span class="fl">2</span>,<span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">plant_02mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="va">plant</span>,<span class="fl">2</span>,<span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>The user specifies the k and m parameters for (mn)-mer
generation.</p>
<p>For example, k = 2 and m = 1 produce the (1,1)-mer, in which one base
is conditioned on the frequency of the base before it. Bases other than
A, C, T, and G were disregarded.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_21mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="va">human</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">plant_21mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mnmer.html">mnmer</a></span><span class="op">(</span><span class="va">plant</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Here, we utilize the (1,1)-mer feature matrices generated by the
mnmer to run an classification using Caret and MLeval.</p>
<p>Caret (<a href="https://topepo.github.io/caret/" class="external-link uri">https://topepo.github.io/caret/</a>) is a library of
functions for building predictive models from R systems. We utilized the
createDataPartition, trainControl, and train functions in this example.
The createDataPartition method splits the feature matrix and creates the
train and test datasets. The trainControl function generates parameters
that further regulate how the train function creates models.</p>
<p>MLeval (<a href="https://cran.r-project.org/web/packages/MLeval/index.html" class="external-link uri">https://cran.r-project.org/web/packages/MLeval/index.html</a>)
is used to plot performance metrics.</p>
<p>To execute the example, enter the following code:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span></span>
<span><span class="co"># Add class information </span></span>
<span><span class="va">classes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">human_21mer</span><span class="op">)</span>, <span class="st">"human.vir"</span><span class="op">)</span></span>
<span><span class="va">featureMatrix_human_21mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">human_21mer</span>,<span class="va">classes</span><span class="op">)</span></span>
<span><span class="va">classes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">plant_21mer</span><span class="op">)</span>, <span class="st">"plant.vir"</span><span class="op">)</span></span>
<span><span class="va">featureMatrix_plant_21mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">plant_21mer</span>,<span class="va">classes</span><span class="op">)</span></span>
<span><span class="va">featureMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">featureMatrix_human_21mer</span>, <span class="va">featureMatrix_plant_21mer</span><span class="op">)</span></span>
<span><span class="va">featureMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">featureMatrix</span>, select <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">seqid</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Machine Learning</span></span>
<span><span class="va">train_index</span> <span class="op">&lt;-</span> <span class="fu">caret</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/caret/man/createDataPartition.html" class="external-link">createDataPartition</a></span><span class="op">(</span><span class="va">featureMatrix</span><span class="op">$</span><span class="va">classes</span>, p<span class="op">=</span><span class="fl">0.8</span>, list<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">train</span> <span class="op">&lt;-</span> <span class="va">featureMatrix</span><span class="op">[</span><span class="va">train_index</span>, <span class="op">]</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">featureMatrix</span><span class="op">[</span><span class="op">-</span><span class="va">train_index</span>, <span class="op">]</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">caret</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/caret/man/trainControl.html" class="external-link">trainControl</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"cv"</span>, </span>
<span>                               summaryFunction<span class="op">=</span><span class="va">twoClassSummary</span>, </span>
<span>                               classProbs<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                               savePredictions <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">roc</span> <span class="op">&lt;-</span> <span class="fu">caret</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/caret/man/train.html" class="external-link">train</a></span><span class="op">(</span><span class="va">classes</span> <span class="op">~</span> <span class="va">.</span>, </span>
<span>                    data<span class="op">=</span><span class="va">train</span>, </span>
<span>                    method<span class="op">=</span><span class="st">"rf"</span>, </span>
<span>                    preProc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span><span class="op">)</span>, </span>
<span>                    trControl<span class="op">=</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">MLeval</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MLeval/man/evalm.html" class="external-link">evalm</a></span><span class="op">(</span><span class="va">roc</span><span class="op">)</span> <span class="co"># Make the ROC plot</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] vctrs_0.6.2       cli_3.6.1         knitr_1.42        rlang_1.1.1      </span></span>
<span><span class="co">##  [5] xfun_0.39         stringi_1.7.12    purrr_1.0.1       textshaping_0.3.6</span></span>
<span><span class="co">##  [9] jsonlite_1.8.4    glue_1.6.2        rprojroot_2.0.3   htmltools_0.5.5  </span></span>
<span><span class="co">## [13] ragg_1.2.5        sass_0.4.5        rmarkdown_2.21    evaluate_0.20    </span></span>
<span><span class="co">## [17] jquerylib_0.1.4   fastmap_1.1.1     yaml_2.3.7        lifecycle_1.0.3  </span></span>
<span><span class="co">## [21] memoise_2.0.1     stringr_1.5.0     compiler_4.3.0    fs_1.6.2         </span></span>
<span><span class="co">## [25] systemfonts_1.0.4 digest_0.6.31     R6_2.5.1          magrittr_2.0.3   </span></span>
<span><span class="co">## [29] bslib_0.4.2       tools_4.3.0       pkgdown_2.0.7     cachem_1.0.8     </span></span>
<span><span class="co">## [33] desc_1.4.2</span></span></code></pre>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Amanda Andrade, Marco Grivet, Otavio Brustolini, Ana Vasconcelos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
